File: native-schema-registry/golang/pkg/gsrserde-go/core/config_test.go
package gsrserde

import (
	"context"
	"errors"
	"net/http"
	"strings"
	"testing"

	awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
	smithymiddleware "github.com/aws/smithy-go/middleware"
	smithyhttp "github.com/aws/smithy-go/transport/http"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// TestLoadConfigFromMapDefaults locks down the defaults Java applies when no
// config keys are present. Anchored to
// common/src/main/java/com/amazonaws/services/schemaregistry/common/configs/GlueSchemaRegistryConfiguration.java
// and AWSSchemaRegistryConstants.java.
func TestLoadConfigFromMapDefaults(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{})
	require.NoError(t, err)

	assert.Equal(t, DefaultRegistryName, cfg.RegistryName)
	assert.Equal(t, DefaultCompatibility, cfg.Compatibility)
	assert.Equal(t, DefaultCompressionType, cfg.CompressionType)
	assert.Equal(t, DefaultCacheTTLMillis, cfg.TimeToLiveMillis)
	assert.Equal(t, DefaultCacheSize, cfg.CacheSize)
	assert.False(t, cfg.SchemaAutoRegistrationEnabled)
	// Description is synthesized from resolved region + registryName per §3.2.
	// With no region and registryName defaulted to "default-registry", the
	// resulting string is "DEFAULT-DESCRIPTION--default-registry" (two dashes
	// — the region segment is empty). Pinned by AC-2 / INV-5.
	assert.Equal(t, "DEFAULT-DESCRIPTION--default-registry", cfg.Description)
	assert.Empty(t, cfg.Endpoint)
	assert.Empty(t, cfg.ProxyURL)
	assert.Empty(t, cfg.UserAgentApp)
	assert.Empty(t, cfg.AssumeRoleArn)
	assert.Empty(t, cfg.SecondaryDeserializer)
	assert.Empty(t, cfg.AvroRecordType)
	assert.Empty(t, cfg.ProtobufMessageType)
	assert.Empty(t, cfg.Tags)
	assert.Empty(t, cfg.Metadata)
}

// Each Java config key gets its own per-key Tier-1 test asserting both the
// parsed Config field AND any side effect on the resulting aws.Config (where
// applicable). Anchored to plan §7 Phase 1 item 5: "one Tier-1 test per key
// verifying the side effect (e.g. userAgentApp reaching aws.Config, region
// honored, etc.)."

func TestConfig_Region(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyRegion: "us-east-2"})
	require.NoError(t, err)
	assert.Equal(t, "us-east-2", cfg.Region)
	// Side effect: aws.Config.Region honors the input.
	assert.Equal(t, "us-east-2", cfg.AWSConfig.Region)
}

func TestConfig_Endpoint(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyEndpoint: "https://glue.local"})
	require.NoError(t, err)
	assert.Equal(t, "https://glue.local", cfg.Endpoint)
}

func TestConfig_ProxyURL_PopulatesHTTPClient(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyProxyURL: "http://proxy.local:8888"})
	require.NoError(t, err)
	assert.Equal(t, "http://proxy.local:8888", cfg.ProxyURL)
	// Side effect: aws.Config.HTTPClient is non-nil (the default chain
	// installs an HTTP client lazily, so a non-nil here means the proxy
	// option ran).
	assert.NotNil(t, cfg.AWSConfig.HTTPClient)
}

func TestConfig_ProxyURL_InvalidErrors(t *testing.T) {
	_, err := LoadConfigFromMap(map[string]string{ConfigKeyProxyURL: "://not a url"})
	require.Error(t, err)
}

func TestConfig_RegistryName(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyRegistryName: "my-registry"})
	require.NoError(t, err)
	assert.Equal(t, "my-registry", cfg.RegistryName)
}

func TestConfig_Compatibility(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyCompatibility: "FULL_ALL"})
	require.NoError(t, err)
	assert.Equal(t, "FULL_ALL", cfg.Compatibility)
}

func TestConfig_Description(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyDescription: "go canary"})
	require.NoError(t, err)
	assert.Equal(t, "go canary", cfg.Description)
}

func TestConfig_SchemaAutoRegistration(t *testing.T) {
	for _, tt := range []struct {
		in   string
		want bool
	}{
		{"true", true},
		{"TRUE", true},
		{"True", true},
		{"false", false},
		{"FALSE", false},
		{"", false},     // omitted ⇒ default false
		{"yes", false},  // Java Boolean.parseBoolean is strict — only "true" is true
		{"1", false},    // same
	} {
		t.Run(tt.in, func(t *testing.T) {
			m := map[string]string{}
			if tt.in != "" {
				m[ConfigKeySchemaAutoRegistration] = tt.in
			}
			cfg, err := LoadConfigFromMap(m)
			require.NoError(t, err)
			assert.Equal(t, tt.want, cfg.SchemaAutoRegistrationEnabled, "input=%q", tt.in)
		})
	}
}

func TestConfig_CompressionType_JavaKeyWins(t *testing.T) {
	// Both keys set; Java's "compression" key wins (the historical Go
	// "compressionType" stays in for back-compat but the canonical name is
	// what Java emits).
	cfg, err := LoadConfigFromMap(map[string]string{
		"compressionType":        "NONE",
		ConfigKeyCompressionType: "ZLIB",
	})
	require.NoError(t, err)
	assert.Equal(t, "ZLIB", cfg.CompressionType)
}

func TestConfig_CompressionType_LegacyGoKey(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{"compressionType": "ZLIB"})
	require.NoError(t, err)
	assert.Equal(t, "ZLIB", cfg.CompressionType)
}

func TestConfig_CacheTTLMillis(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyCacheTTLMillis: "1234"})
	require.NoError(t, err)
	assert.Equal(t, int64(1234), cfg.TimeToLiveMillis)
}

// TestConfig_CacheTTLMillis_NonNumericErrors locks down the §3.1 semantic
// flip: non-numeric input no longer falls back to default; it surfaces as
// ErrInvalidCacheTTL. Renamed from
// TestConfig_CacheTTLMillis_NonNumericFallsBackToDefault per INV-11.
func TestConfig_CacheTTLMillis_NonNumericErrors(t *testing.T) {
	_, err := LoadConfigFromMap(map[string]string{ConfigKeyCacheTTLMillis: "not-a-number"})
	require.Error(t, err)
	assert.True(t, errors.Is(err, ErrInvalidCacheTTL), "want ErrInvalidCacheTTL, got %v", err)

	// Empty input still picks the default.
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyCacheTTLMillis: ""})
	require.NoError(t, err)
	assert.Equal(t, DefaultCacheTTLMillis, cfg.TimeToLiveMillis)
}

func TestConfig_CacheSize(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyCacheSize: "42"})
	require.NoError(t, err)
	assert.Equal(t, 42, cfg.CacheSize)
}

func TestConfig_AvroRecordType(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyAvroRecordType: "GENERIC_RECORD"})
	require.NoError(t, err)
	assert.Equal(t, "GENERIC_RECORD", cfg.AvroRecordType)
}

func TestConfig_ProtobufMessageType(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyProtobufMessageType: "DYNAMIC_MESSAGE"})
	require.NoError(t, err)
	assert.Equal(t, "DYNAMIC_MESSAGE", cfg.ProtobufMessageType)
}

func TestConfig_SecondaryDeserializer(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeySecondaryDeserializer: "com.example.Foo"})
	require.NoError(t, err)
	assert.Equal(t, "com.example.Foo", cfg.SecondaryDeserializer)
}

func TestConfig_UserAgentApp(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyUserAgentApp: "my-service"})
	require.NoError(t, err)
	assert.Equal(t, "my-service", cfg.UserAgentApp)
	// Side effect: APIOptions is non-empty (the user-agent middleware was
	// appended). We don't trigger a real Glue call here so we can't read the
	// final User-Agent header; the presence of an APIOption is the visible
	// surface in unit-test land.
	assert.NotEmpty(t, cfg.AWSConfig.APIOptions)
}

func TestConfig_AssumeRoleArn_InstallsCredentialsChain(t *testing.T) {
	// We can't reach AWS here, so we assert that the credentials chain
	// installed a non-nil provider; constructing one is the side effect.
	cfg, err := LoadConfigFromMap(map[string]string{
		ConfigKeyAssumeRoleArn:         "arn:aws:iam::123456789012:role/go-test",
		ConfigKeyAssumeRoleSessionName: "session-1",
	})
	require.NoError(t, err)
	assert.Equal(t, "arn:aws:iam::123456789012:role/go-test", cfg.AssumeRoleArn)
	assert.Equal(t, "session-1", cfg.AssumeRoleSessionName)
	assert.NotNil(t, cfg.AWSConfig.Credentials)
}

func TestConfig_AssumeRoleArn_DefaultSessionName(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{
		ConfigKeyAssumeRoleArn: "arn:aws:iam::123456789012:role/go-test",
	})
	require.NoError(t, err)
	// AssumeRoleSessionName field reflects raw input ("" when absent), but
	// the credentials provider falls back to DefaultAssumeRoleSession.
	assert.Empty(t, cfg.AssumeRoleSessionName)
}

func TestConfig_SchemaNameGenerationClass(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeySchemaNameGenerationClass: "com.example.MyStrategy"})
	require.NoError(t, err)
	assert.Equal(t, "com.example.MyStrategy", cfg.SchemaNameGenerationClass)
}

func TestConfig_Tags(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{
		"tags.env":  "beta",
		"tags.team": "schema-registry",
	})
	require.NoError(t, err)
	assert.Equal(t, map[string]string{
		"env":  "beta",
		"team": "schema-registry",
	}, cfg.Tags)
}

func TestConfig_Metadata(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{
		"metadata.commit":  "abc123",
		"metadata.version": "1.0.0",
	})
	require.NoError(t, err)
	assert.Equal(t, map[string]string{
		"commit":  "abc123",
		"version": "1.0.0",
	}, cfg.Metadata)
}

// TestConfig_CompressionType_InvalidErrors locks down §3.1 — invalid
// compression values fail-fast with ErrInvalidCompressionType. Plan §7 pins
// the "GZip" case specifically.
func TestConfig_CompressionType_InvalidErrors(t *testing.T) {
	for _, in := range []string{"GZip", "snappy"} {
		t.Run(in, func(t *testing.T) {
			_, err := LoadConfigFromMap(map[string]string{ConfigKeyCompressionType: in})
			require.Error(t, err)
			assert.True(t, errors.Is(err, ErrInvalidCompressionType), "want ErrInvalidCompressionType, got %v", err)
		})
	}
}

// TestConfig_Compatibility_InvalidErrors locks down §3.1 — the Java
// Compatibility enum is CASE-EXACT, so lowercase variants and typos MUST be
// rejected. See GlueSchemaRegistryConfiguration.java:173-178.
func TestConfig_Compatibility_InvalidErrors(t *testing.T) {
	for _, in := range []string{"FORWARDS", "backward"} {
		t.Run(in, func(t *testing.T) {
			_, err := LoadConfigFromMap(map[string]string{ConfigKeyCompatibility: in})
			require.Error(t, err)
			assert.True(t, errors.Is(err, ErrInvalidCompatibility), "want ErrInvalidCompatibility, got %v", err)
		})
	}
}

// TestConfig_CacheSize_NonNumericErrors locks down §3.1 — non-numeric
// cacheSize input surfaces as ErrInvalidCacheSize; empty input picks default.
func TestConfig_CacheSize_NonNumericErrors(t *testing.T) {
	_, err := LoadConfigFromMap(map[string]string{ConfigKeyCacheSize: "NaN"})
	require.Error(t, err)
	assert.True(t, errors.Is(err, ErrInvalidCacheSize), "want ErrInvalidCacheSize, got %v", err)

	cfg, err := LoadConfigFromMap(map[string]string{ConfigKeyCacheSize: ""})
	require.NoError(t, err)
	assert.Equal(t, DefaultCacheSize, cfg.CacheSize)
}

// TestConfig_DescriptionDefault_SynthesizedFromRegionAndRegistry locks down
// AC-2 / INV-5 / S-8: when the description key is absent or empty,
// LoadConfigFromMap synthesizes "DEFAULT-DESCRIPTION-<region>-<registryName>"
// using the RESOLVED region and registry name (post-default fallback). Mirrors
// Java GlueSchemaRegistryConfiguration.java:343-352. The "DEFAULT-DESCRIPTION"
// prefix is exact — no casing change. Empty region collapses to two dashes.
func TestConfig_DescriptionDefault_SynthesizedFromRegionAndRegistry(t *testing.T) {
	t.Run("region and registry both set", func(t *testing.T) {
		cfg, err := LoadConfigFromMap(map[string]string{
			ConfigKeyRegion:       "us-east-2",
			ConfigKeyRegistryName: "my-registry",
		})
		require.NoError(t, err)
		assert.Equal(t, "DEFAULT-DESCRIPTION-us-east-2-my-registry", cfg.Description)
	})

	t.Run("region set, registry defaulted", func(t *testing.T) {
		cfg, err := LoadConfigFromMap(map[string]string{
			ConfigKeyRegion: "us-east-2",
		})
		require.NoError(t, err)
		// Registry name resolves to DefaultRegistryName; description sees the
		// post-default value, not the raw absent key.
		assert.Equal(t, DefaultRegistryName, cfg.RegistryName)
		assert.Equal(t, "DEFAULT-DESCRIPTION-us-east-2-default-registry", cfg.Description)
	})

	t.Run("region and registry absent", func(t *testing.T) {
		cfg, err := LoadConfigFromMap(map[string]string{})
		require.NoError(t, err)
		// Empty region segment collapses to two dashes — Java parity.
		assert.Equal(t, "DEFAULT-DESCRIPTION--default-registry", cfg.Description)
	})

	t.Run("explicit description passes through unchanged", func(t *testing.T) {
		cfg, err := LoadConfigFromMap(map[string]string{
			ConfigKeyDescription: "user supplied",
			ConfigKeyRegion:      "us-east-2",
		})
		require.NoError(t, err)
		assert.Equal(t, "user supplied", cfg.Description)
	})
}

// TestConfig_Validators_EmptyInputsStillSucceed locks AC-1-positive: when all
// four validated keys are absent / empty, LoadConfigFromMap succeeds and the
// defaults are honored.
func TestConfig_Validators_EmptyInputsStillSucceed(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{
		ConfigKeyCompressionType: "",
		ConfigKeyCompatibility:   "",
		ConfigKeyCacheTTLMillis:  "",
		ConfigKeyCacheSize:       "",
	})
	require.NoError(t, err)
	assert.Equal(t, DefaultCompressionType, cfg.CompressionType)
	assert.Equal(t, DefaultCompatibility, cfg.Compatibility)
	assert.Equal(t, DefaultCacheTTLMillis, cfg.TimeToLiveMillis)
	assert.Equal(t, DefaultCacheSize, cfg.CacheSize)
}

// Sanity check that the constants exposed by core/config.go are the
// string values Java emits, not Go-idiomatic snake_case. Java callers
// targeting the same config keys must get the same behavior.
func TestConfigKeys_AreJavaIdentical(t *testing.T) {
	pairs := map[string]string{
		"region":                        ConfigKeyRegion,
		"endpoint":                      ConfigKeyEndpoint,
		"proxyUrl":                      ConfigKeyProxyURL,
		"registry.name":                 ConfigKeyRegistryName,
		"compatibility":                 ConfigKeyCompatibility,
		"description":                   ConfigKeyDescription,
		"schemaAutoRegistrationEnabled": ConfigKeySchemaAutoRegistration,
		"compression":                   ConfigKeyCompressionType,
		"timeToLiveMillis":              ConfigKeyCacheTTLMillis,
		"cacheSize":                     ConfigKeyCacheSize,
		"avroRecordType":                ConfigKeyAvroRecordType,
		"protobufMessageType":           ConfigKeyProtobufMessageType,
		"secondaryDeserializer":         ConfigKeySecondaryDeserializer,
		"userAgentApp":                  ConfigKeyUserAgentApp,
		"assumeRoleArn":                 ConfigKeyAssumeRoleArn,
		"assumeRoleSessionName":         ConfigKeyAssumeRoleSessionName,
		"schemaNameGenerationClass":     ConfigKeySchemaNameGenerationClass,
	}
	for want, got := range pairs {
		assert.Equal(t, want, got, "config-key constant must match the Java constant string")
	}
}

// TestConfig_UserAgentApp_DefaultAppliedWhenAbsent locks down AC-3 / INV-3:
// when `userAgentApp` is absent, the raw Config.UserAgentApp stays empty,
// Config.EffectiveUserAgentApp resolves to "default", and the User-Agent
// middleware is still installed on AWSConfig.APIOptions (Java always-on
// parity per GlueSchemaRegistryConfiguration.java:66).
func TestConfig_UserAgentApp_DefaultAppliedWhenAbsent(t *testing.T) {
	cfg, err := LoadConfigFromMap(map[string]string{})
	require.NoError(t, err)

	// INV-3-raw: raw field reflects raw input ("" when absent).
	assert.Empty(t, cfg.UserAgentApp)
	// INV-3-effective: resolved field carries "default" when raw is empty.
	assert.Equal(t, DefaultUserAgentApp, cfg.EffectiveUserAgentApp)
	// AC-3: middleware ALWAYS installed (Java parity).
	assert.NotEmpty(t, cfg.AWSConfig.APIOptions)
}

// TestConfig_UserAgentApp_HeaderEmitsResolvedValue exercises the installed
// API options against a synthetic smithy middleware stack. The captured
// User-Agent header MUST contain "glue-schema-registry-go/<resolved>" for
// both the default (key absent) and explicit (key set) cases. Pins AC-3 /
// S-3 / S-4.
func TestConfig_UserAgentApp_HeaderEmitsResolvedValue(t *testing.T) {
	for _, tc := range []struct {
		name     string
		cfgMap   map[string]string
		expected string
	}{
		{
			name:     "default applied when key absent",
			cfgMap:   map[string]string{},
			expected: "glue-schema-registry-go/default",
		},
		{
			name:     "explicit value flows through",
			cfgMap:   map[string]string{ConfigKeyUserAgentApp: "my-service"},
			expected: "glue-schema-registry-go/my-service",
		},
	} {
		t.Run(tc.name, func(t *testing.T) {
			cfg, err := LoadConfigFromMap(tc.cfgMap)
			require.NoError(t, err)
			require.NotEmpty(t, cfg.AWSConfig.APIOptions, "user-agent middleware must be installed")

			// Build a synthetic smithy stack, apply each captured APIOption,
			// then run the stack against a no-op handler. The
			// awsmiddleware.RequestUserAgent middleware writes the
			// User-Agent header in its Build step; we inspect it via an
			// inspector middleware appended after Build.
			stack := smithymiddleware.NewStack("ua-test", smithyhttp.NewStackRequest)
			for _, opt := range cfg.AWSConfig.APIOptions {
				require.NoError(t, opt(stack))
			}
			// Force-install the SDK's RequestUserAgent middleware on the
			// build step (the AddUserAgentKey option only mutates an
			// existing middleware; if none was previously added the option
			// adds one, which is what we need here).
			require.NoError(t, awsmiddleware.AddRequestUserAgentMiddleware(stack))

			var captured http.Header
			require.NoError(t, stack.Build.Add(
				smithymiddleware.BuildMiddlewareFunc("ua-inspector",
					func(ctx context.Context, in smithymiddleware.BuildInput, next smithymiddleware.BuildHandler) (smithymiddleware.BuildOutput, smithymiddleware.Metadata, error) {
						req := in.Request.(*smithyhttp.Request)
						captured = req.Header.Clone()
						return next.HandleBuild(ctx, in)
					}),
				smithymiddleware.After,
			))

			handler := smithymiddleware.DecorateHandler(
				smithymiddleware.HandlerFunc(func(ctx context.Context, input interface{}) (interface{}, smithymiddleware.Metadata, error) {
					return nil, smithymiddleware.Metadata{}, nil
				}),
				stack,
			)
			_, _, err = handler.Handle(context.Background(), nil)
			require.NoError(t, err)

			ua := captured.Get("User-Agent")
			require.NotEmpty(t, ua, "User-Agent header must be populated by the middleware")
			assert.True(t, strings.Contains(ua, tc.expected),
				"User-Agent %q must contain %q", ua, tc.expected)
		})
	}
}

File: multilang-schema-registry/typescript/packages/core/src/config/config.test.ts
import { describe, expect, it } from "vitest";

import {
  GsrInvalidAvroRecordTypeError,
  GsrInvalidCacheSizeError,
  GsrInvalidCacheTtlError,
  GsrInvalidCompatibilityError,
  GsrInvalidCompressionTypeError,
  GsrInvalidProtobufMessageTypeError,
} from "../errors/gsr-errors.js";

import {
  DEFAULT_ASSUME_ROLE_SESSION,
  DEFAULT_CACHE_SIZE,
  DEFAULT_CACHE_TTL_MILLIS,
  DEFAULT_COMPATIBILITY,
  DEFAULT_COMPRESSION_TYPE,
  DEFAULT_MAX_ATTEMPTS,
  DEFAULT_REGION,
  DEFAULT_REGISTRY_NAME,
  DEFAULT_RETRY_MODE,
  DEFAULT_USER_AGENT_APP,
  parseConfig,
} from "./config.js";

describe("parseConfig — defaults on an empty config map", () => {
  it("applies every Java/Go-parity default and synthesizes the description", () => {
    const cfg = parseConfig({});

    expect(cfg.region).toBe(DEFAULT_REGION);
    expect(cfg.registryName).toBe(DEFAULT_REGISTRY_NAME);
    expect(cfg.compatibility).toBe(DEFAULT_COMPATIBILITY);
    expect(cfg.compressionType).toBe(DEFAULT_COMPRESSION_TYPE);
    expect(cfg.timeToLiveMillis).toBe(DEFAULT_CACHE_TTL_MILLIS);
    expect(cfg.cacheSize).toBe(DEFAULT_CACHE_SIZE);
    expect(cfg.effectiveUserAgentApp).toBe(DEFAULT_USER_AGENT_APP);
    expect(cfg.retryMode).toBe(DEFAULT_RETRY_MODE);
    expect(cfg.maxAttempts).toBe(DEFAULT_MAX_ATTEMPTS);
    expect(cfg.schemaAutoRegistrationEnabled).toBe(false);

    // Synthesized description reflects post-default region + registryName.
    expect(cfg.description).toBe(
      `DEFAULT-DESCRIPTION-${DEFAULT_REGION}-${DEFAULT_REGISTRY_NAME}`,
    );

    // Optional keys are absent on the default surface.
    expect(cfg.endpoint).toBeUndefined();
    expect(cfg.avroRecordType).toBeUndefined();
    expect(cfg.protobufMessageType).toBeUndefined();
    expect(cfg.avroReaderSchema).toBeUndefined();
    expect(cfg.schemaNameGenerationClass).toBeUndefined();
    expect(cfg.assumeRoleArn).toBeUndefined();

    // Raw userAgentApp is preserved as "" so callers can distinguish default
    // from explicit; the resolved value on effectiveUserAgentApp is "default".
    expect(cfg.userAgentApp).toBe("");
    // Session name is empty when no ARN is configured (Go dormant-path parity).
    expect(cfg.assumeRoleSessionName).toBe("");

    // tags / metadata prefix-maps are always present, empty on default input.
    expect(cfg.tags).toEqual({});
    expect(cfg.metadata).toEqual({});
  });
});

describe("parseConfig — compatibility validation (case-exact, 8 modes)", () => {
  const validModes = [
    "NONE",
    "DISABLED",
    "BACKWARD",
    "BACKWARD_ALL",
    "FORWARD",
    "FORWARD_ALL",
    "FULL",
    "FULL_ALL",
  ] as const;

  it.each(validModes)("accepts %s case-exact", (mode) => {
    expect(parseConfig({ compatibility: mode }).compatibility).toBe(mode);
  });

  it.each(["backward", "FORWARDS", "backward_all", ""])(
    "rejects %j",
    (bad) => {
      expect(() => parseConfig({ compatibility: bad })).toThrow(
        GsrInvalidCompatibilityError,
      );
    },
  );
});

describe("parseConfig — compression validation (explicit-only)", () => {
  it("accepts NONE and ZLIB", () => {
    expect(parseConfig({ compression: "NONE" }).compressionType).toBe("NONE");
    expect(parseConfig({ compression: "ZLIB" }).compressionType).toBe("ZLIB");
  });

  it("throws GsrInvalidCompressionTypeError on explicit out-of-set value", () => {
    expect(() => parseConfig({ compression: "GZIP" })).toThrow(
      GsrInvalidCompressionTypeError,
    );
  });

  it("falls through to NONE default when the compression key is absent (no throw)", () => {
    const cfg = parseConfig({});
    expect(cfg.compressionType).toBe("NONE");
  });

  it("also accepts the legacy compressionType key, with the Java key winning on collision", () => {
    const legacyOnly = parseConfig({ compressionType: "ZLIB" });
    expect(legacyOnly.compressionType).toBe("ZLIB");

    const bothSet = parseConfig({ compressionType: "NONE", compression: "ZLIB" });
    expect(bothSet.compressionType).toBe("ZLIB");
  });
});

describe("parseConfig — integer validation on cache keys", () => {
  it("parses valid integer strings", () => {
    const cfg = parseConfig({ timeToLiveMillis: "60000", cacheSize: "500" });
    expect(cfg.timeToLiveMillis).toBe(60_000);
    expect(cfg.cacheSize).toBe(500);
  });

  it.each(["notanumber", "1.5", "1e6", "10abc", " 10 ", ""])(
    "throws GsrInvalidCacheTtlError on non-integer timeToLiveMillis %j (except empty which falls back)",
    (raw) => {
      if (raw === "") {
        expect(parseConfig({ timeToLiveMillis: raw }).timeToLiveMillis).toBe(
          DEFAULT_CACHE_TTL_MILLIS,
        );
      } else {
        expect(() => parseConfig({ timeToLiveMillis: raw })).toThrow(
          GsrInvalidCacheTtlError,
        );
      }
    },
  );

  it("throws GsrInvalidCacheSizeError on non-integer cacheSize", () => {
    expect(() => parseConfig({ cacheSize: "abc" })).toThrow(
      GsrInvalidCacheSizeError,
    );
    expect(() => parseConfig({ cacheSize: "1.5" })).toThrow(
      GsrInvalidCacheSizeError,
    );
  });
});

describe("parseConfig — avroRecordType / protobufMessageType (case-exact)", () => {
  it("accepts the case-exact avroRecordType values", () => {
    expect(parseConfig({ avroRecordType: "GENERIC_RECORD" }).avroRecordType).toBe(
      "GENERIC_RECORD",
    );
    expect(parseConfig({ avroRecordType: "SPECIFIC_RECORD" }).avroRecordType).toBe(
      "SPECIFIC_RECORD",
    );
  });

  it("throws GsrInvalidAvroRecordTypeError on lowercase or typo variants", () => {
    expect(() => parseConfig({ avroRecordType: "generic_record" })).toThrow(
      GsrInvalidAvroRecordTypeError,
    );
    expect(() => parseConfig({ avroRecordType: "GenericRecord" })).toThrow(
      GsrInvalidAvroRecordTypeError,
    );
  });

  it("accepts the case-exact protobufMessageType values", () => {
    expect(parseConfig({ protobufMessageType: "POJO" }).protobufMessageType).toBe(
      "POJO",
    );
    expect(
      parseConfig({ protobufMessageType: "DYNAMIC_MESSAGE" }).protobufMessageType,
    ).toBe("DYNAMIC_MESSAGE");
  });

  it("throws GsrInvalidProtobufMessageTypeError on lowercase or typo variants", () => {
    expect(() => parseConfig({ protobufMessageType: "pojo" })).toThrow(
      GsrInvalidProtobufMessageTypeError,
    );
    expect(() => parseConfig({ protobufMessageType: "dynamic" })).toThrow(
      GsrInvalidProtobufMessageTypeError,
    );
  });

  it("leaves both fields undefined when the keys are absent", () => {
    const cfg = parseConfig({});
    expect(cfg.avroRecordType).toBeUndefined();
    expect(cfg.protobufMessageType).toBeUndefined();
  });
});

describe("parseConfig — accept-and-ignore + accept-and-store keys", () => {
  it("does not throw on secondaryDeserializer (accept-and-ignore, removed)", () => {
    const cfg = parseConfig({ secondaryDeserializer: "com.example.Legacy" });
    // No public field surfaces the value.
    expect(cfg).not.toHaveProperty("secondaryDeserializer");
  });

  it("does not throw on proxyUrl (accept-and-ignore, deferred to real-AWS wiring)", () => {
    const cfg = parseConfig({ proxyUrl: "http://proxy.internal:3128" });
    expect(cfg).not.toHaveProperty("proxyUrl");
  });

  it("accepts avroReaderSchema and stores it unvalidated (deliberate Go-parity divergence)", () => {
    // A syntactically-broken value is deliberately NOT rejected at parse time —
    // core cannot import an Avro parser. Validation is deferred to projection.
    const broken = "{ not valid avro json";
    const cfg = parseConfig({ avroReaderSchema: broken });
    expect(cfg.avroReaderSchema).toBe(broken);
  });

  it("accepts schemaNameGenerationClass and stores the raw string", () => {
    const cfg = parseConfig({
      schemaNameGenerationClass: "com.example.MyStrategy",
    });
    expect(cfg.schemaNameGenerationClass).toBe("com.example.MyStrategy");
  });
});

describe("parseConfig — user-agent resolution", () => {
  it("preserves the raw userAgentApp and resolves the effective token", () => {
    const cfg = parseConfig({ userAgentApp: "myapp" });
    expect(cfg.userAgentApp).toBe("myapp");
    expect(cfg.effectiveUserAgentApp).toBe("myapp");
  });

  it("falls back to the default when userAgentApp is empty", () => {
    const cfg = parseConfig({ userAgentApp: "" });
    expect(cfg.userAgentApp).toBe("");
    expect(cfg.effectiveUserAgentApp).toBe(DEFAULT_USER_AGENT_APP);
  });
});

describe("parseConfig — AssumeRole session name resolution", () => {
  it("resolves to the caller-provided value when both keys are set", () => {
    const cfg = parseConfig({
      assumeRoleArn: "arn:aws:iam::123456789012:role/gsr",
      assumeRoleSessionName: "override",
    });
    expect(cfg.assumeRoleArn).toBe("arn:aws:iam::123456789012:role/gsr");
    expect(cfg.assumeRoleSessionName).toBe("override");
  });

  it("resolves to the default when ARN is set but session name is absent", () => {
    const cfg = parseConfig({
      assumeRoleArn: "arn:aws:iam::123456789012:role/gsr",
    });
    expect(cfg.assumeRoleSessionName).toBe(DEFAULT_ASSUME_ROLE_SESSION);
  });

  it("leaves session name empty when no ARN is configured (dormant path)", () => {
    const cfg = parseConfig({});
    expect(cfg.assumeRoleSessionName).toBe("");
    expect(cfg.assumeRoleArn).toBeUndefined();
  });
});

describe("parseConfig — tag and metadata prefix-expansion", () => {
  it("expands tags.<k>=<v> and metadata.<k>=<v> entries into their maps", () => {
    const cfg = parseConfig({
      "tags.owner": "team-gsr",
      "tags.env": "beta",
      "metadata.build": "abc123",
      "metadata.git": "sha:def",
      // Non-prefixed key MUST NOT bleed into either map.
      "registry.name": "custom",
    });
    expect(cfg.tags).toEqual({ owner: "team-gsr", env: "beta" });
    expect(cfg.metadata).toEqual({ build: "abc123", git: "sha:def" });
    expect(cfg.registryName).toBe("custom");
  });

  it("returns empty maps when no prefixed keys are present", () => {
    const cfg = parseConfig({ region: "eu-west-1" });
    expect(cfg.tags).toEqual({});
    expect(cfg.metadata).toEqual({});
  });
});

describe("parseConfig — schemaAutoRegistrationEnabled (lenient bool)", () => {
  it.each(["true", "True", "TRUE"])("treats %j as true", (raw) => {
    expect(parseConfig({ schemaAutoRegistrationEnabled: raw }).schemaAutoRegistrationEnabled).toBe(
      true,
    );
  });

  it.each(["false", "yes", "1", "", "no"])("treats %j as false", (raw) => {
    expect(parseConfig({ schemaAutoRegistrationEnabled: raw }).schemaAutoRegistrationEnabled).toBe(
      false,
    );
  });
});

describe("parseConfig — retry surface (TS-additive)", () => {
  it("accepts the two documented retry modes", () => {
    expect(parseConfig({ retryMode: "standard" }).retryMode).toBe("standard");
    expect(parseConfig({ retryMode: "adaptive" }).retryMode).toBe("adaptive");
  });

  it("silently falls back to the default on out-of-set retryMode values (Go has no key)", () => {
    expect(parseConfig({ retryMode: "aggressive" }).retryMode).toBe(
      DEFAULT_RETRY_MODE,
    );
  });

  it("parses maxAttempts as an integer, falling back on non-integer input", () => {
    expect(parseConfig({ maxAttempts: "5" }).maxAttempts).toBe(5);
    expect(parseConfig({ maxAttempts: "1.5" }).maxAttempts).toBe(
      DEFAULT_MAX_ATTEMPTS,
    );
    expect(parseConfig({ maxAttempts: "abc" }).maxAttempts).toBe(
      DEFAULT_MAX_ATTEMPTS,
    );
  });
});

describe("parseConfig — region / endpoint / description resolution", () => {
  it("uses the caller-provided region and endpoint when set", () => {
    const cfg = parseConfig({
      region: "eu-west-1",
      endpoint: "https://glue.local",
    });
    expect(cfg.region).toBe("eu-west-1");
    expect(cfg.endpoint).toBe("https://glue.local");
  });

  it("applies the region default when the key is empty", () => {
    const cfg = parseConfig({ region: "" });
    expect(cfg.region).toBe(DEFAULT_REGION);
  });

  it("respects an explicit description over the synthesized default", () => {
    const cfg = parseConfig({ description: "custom-desc" });
    expect(cfg.description).toBe("custom-desc");
  });

  it("re-synthesizes description after applied registry-name default", () => {
    const cfg = parseConfig({ region: "eu-west-1", "registry.name": "prod" });
    expect(cfg.description).toBe("DEFAULT-DESCRIPTION-eu-west-1-prod");
  });
});

describe("parseConfig — purity", () => {
  it("does not mutate the caller's config map", () => {
    const original: Record<string, string> = {
      region: "eu-west-1",
      "tags.env": "beta",
    };
    const before = JSON.stringify(original);
    parseConfig(original);
    expect(JSON.stringify(original)).toBe(before);
  });
});
